infrastructure/src/main/java/org/corfudb/infrastructure/management/FailureDetector.java

```diff
@@ -283,103 +279,43 @@ private NodeState getLocalNodeState(Set<String> allServers, Set<String> allConne
      * - build poll report
      *
      * @param allServers       all servers in the cluster
-     * @param clientRouters           client clientRouters
+     * @param clientRouters    client clientRouters
      * @param epoch            current epoch
      * @param sequencerMetrics metrics
      * @param layout           current layout
      * @return a poll report
      */
     private PollReport pollIteration(
-            Set<String> allServers, Map<String, IClientRouter> clientRouters, long epoch, SequencerMetrics sequencerMetrics,
-            Layout layout) {
+            Set<String> allServers, Map<String, IClientRouter> clientRouters, long epoch,
+            SequencerMetrics sequencerMetrics, Layout layout) {
 
         log.trace("Poll iteration. Epoch: {}", epoch);
 
-        Map<String, CompletableFuture<NodeState>> stateFuture = pollAsync(allServers, clientRouters, epoch);
+        ClusterStateCollector clusterCollector = ClusterStateCollector.builder()
+                .localEndpoint(localEndpoint)
+                .timeout(period)
+                .clusterState(pollAsync(allServers, clientRouters, epoch))
+                .heartbeatCounter(heartbeatCounter)
```

Comment:
The `heartbeatCounter` should be incremented every round and not every iteration.
The `heartbeatCounter` is sort of a version number to your cluster state which you update.

Reply:
I copy pasted this logic from the previous version. Not sure if we need to change it in this PR.

Reply:
Ok. But this needs cleanup too.
You can move the incrementAndGet heartbeatcounter here rather than in the aggregator.
That will be easier to understand.

Reply:
let's do it in another pr